node/src/android/OpenT2TClient/src/main/jni/JniUtils.h

```diff
@@ -0,0 +1,98 @@
+
+void resolvePromise(JNIEnv* env, jobject promise, jobject result)
+{
+    OpenT2T::LogTrace("resolvePromise()");
+
+    jclass promiseClass = env->GetObjectClass(promise);
+    jmethodID resolveMethod = env->GetMethodID(
+            promiseClass, "resolve", "(Ljava/lang/Object;)V");
+    env->CallVoidMethod(promise, resolveMethod, result);
+
+    if (env->ExceptionOccurred())
+    {
+        OpenT2T::LogError("Promise resolve threw exception");
+        env->ExceptionClear();
+    }
+}
+
+void rejectPromise(JNIEnv* env, jobject promise, jthrowable ex)
+{
+    OpenT2T::LogTrace("rejectPromise()");
+
+    jclass promiseClass = env->GetObjectClass(promise);
+    jmethodID rejectMethod = env->GetMethodID(
+            promiseClass, "reject", "(Ljava/lang/Exception;)V");
+    env->CallVoidMethod(promise, rejectMethod, ex);
+
+    if (env->ExceptionOccurred())
+    {
+        OpenT2T::LogError("Promise reject threw exception");
+        env->ExceptionClear();
+    }
+}
+
+jthrowable newJavaException(JNIEnv* env, const char* exceptionClassName, const char* message)
+{
+    OpenT2T::LogTrace("newJavaException(\"%s\", \"%s\")",
+            exceptionClassName, (message != nullptr ? message : ""));
+
+    jthrowable javaException;
+    jclass exceptionClass = env->FindClass(exceptionClassName);
+    if (message == nullptr)
+    {
+        jmethodID exceptionConstructor =
+            env->GetMethodID(exceptionClass, "<init>", "()V");
+        javaException = static_cast<jthrowable>(
+            env->NewObject(exceptionClass, exceptionConstructor));
+    }
+    else
+    {
+        jmethodID exceptionConstructor =
+            env->GetMethodID(exceptionClass, "<init>", "(java/lang/String)V");
+        jstring messageString = env->NewStringUTF(message);
```

Comment:
No... messageString (like the other j* local variables here) is a JNI local reference, so it will be released automatically after the function returns.

---
In reply to: [71086834](https://github.com/openT2T/client/pull/1#discussion_r71086834) [](ancestors = 71086834)